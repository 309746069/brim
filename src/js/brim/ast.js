/* @flow */
export default function(ast: Object) {
  return {
    groupByKeys() {
      let g = this.proc("GroupByProc")
      return g ? g.keys : []
    },
    proc(name: string) {
      return getProcs(ast).find((p) => p.op === name)
    }
  }
}

function getProcs(ast) {
  if (!ast || !ast.proc) return []
  let list = []
  collectProcs(ast.proc, list)
  return list
}

function collectProcs(proc, list) {
  if (COMPOUND_PROCS.includes(proc.op)) {
    for (let p of proc.procs) collectProcs(p, list)
  } else {
    list.push(proc)
  }
}

export const HEAD_PROC = "HeadProc"
export const TAIL_PROC = "TailProc"
export const SORT_PROC = "SortProc"
export const FILTER_PROC = "FilterProc"
export const PARALLEL_PROC = "ParallelProc"
export const SEQUENTIAL_PROC = "SequentialProc"
export const SOURCE_PROC = "SourceProc"
export const TUPLE_PROCS = [
  SOURCE_PROC,
  HEAD_PROC,
  TAIL_PROC,
  SORT_PROC,
  FILTER_PROC,
  SEQUENTIAL_PROC
]
export const COMPOUND_PROCS = [PARALLEL_PROC, SEQUENTIAL_PROC]
