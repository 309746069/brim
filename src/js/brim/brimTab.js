/* @flow */

import type {TabState} from "../state/Tab/types"
import lib from "../lib"

export default function(tab: TabState) {
  return {
    title() {
      return lib
        .compact([tab.search.space || "No Space", tab.searchBar.previous])
        .join(": ")
    }
  }
}
