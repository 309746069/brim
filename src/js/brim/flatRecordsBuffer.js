/* @flow */
import type {FieldData} from "../types/records"
import {isArray, isString} from "../lib/is"

type Col = {name: string, type: string | Descriptor}
type Descriptor = Col[]
type Value = null | string | Value[]

type NestedRecord = {
  id: number,
  type?: Descriptor,
  values: Value[]
}

export default function flatRecordBuffers() {
  let channels = mapArrays()
  let types = {}

  return {
    add(chanId: string, records: NestedRecord[]) {
      channels.append(
        chanId,
        records.map((r) => {
          if (r.type) types[r.id] = r.type
          let record = zipFields(types[r.id], r.values)
          return flattenRecord(record)
        })
      )
    },

    records(chanId: string = "0") {
      return channels.get(chanId)
    },

    empty(chanId: string = "0") {
      return this.records(chanId).length === 0
    },

    columns() {
      return Object.keys(types).reduce((all, id) => {
        all[id] = flattenType(types[id])
        return all
      }, {})
    },

    clearRecords(chanId: string = "0") {
      channels.clear(chanId)
    }
  }
}

function mapArrays() {
  let m = new Map<string, *[]>()

  return {
    append(key, value) {
      let array = this.get(key)
      this.set(key, array.concat(value))
    },
    get(key) {
      return m.get(key) || []
    },
    set(key, val) {
      return m.set(key, val)
    },
    clear(key) {
      return m.set(key, [])
    }
  }
}

function zipFields(types, values): FieldData[] {
  // $FlowFixMe
  return values.map<FieldData>((value, index) => {
    let {name, type} = types[index]
    return isString(type)
      ? {name, type, value}
      : {name, type: "record", value: zipFields(type, value)}
  })
}

function flattenType(descriptor, prefix = "") {
  return descriptor.reduce((flat, {name, type}) => {
    let cols = isArray(type)
      ? flattenType(type, `${name}.`)
      : [{name: prefix + name, type}]

    return flat.concat(cols)
  }, [])
}

function flattenRecord(record, prefix) {
  // $FlowFixMe
  return record.reduce(
    (array, field) => array.concat(flatFields(field, prefix)),
    []
  )
}

function flatFields({name, value, type}, prefix = "") {
  return type === "record"
    ? flattenRecord(value, `${name}.`)
    : [{name: prefix + name, type, value}]
}
