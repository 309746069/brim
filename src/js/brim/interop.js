/* @flow */
import type {$Record} from "../types/records"
import Log from "../models/Log"

export default {
  recordToLog(record: $Record) {
    return new Log(record.values(), record.columns())
  }
}
