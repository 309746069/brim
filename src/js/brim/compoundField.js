/* @flow */
import brim, {type $CompoundField} from "./"

export const COMPOUND_FIELD_RGX = /^(set|vector)\[(\w+)\]$/
type $T = $CompoundField

function compoundField(name: string, type: string, value: string): $T {
  let match = type.match(COMPOUND_FIELD_RGX)
  if (!match) throw new Error("Not compound type: " + type)
  let [_, container, itemType] = match
  let items = value.split(",")

  return {
    name,
    container,
    itemType,
    length: items.length,
    items() {
      return items.map((_, index) => this.item(index))
    },
    item(index: number) {
      if (items[index]) {
        return brim.field(name, itemType, items[index])
      } else {
        return null
      }
    }
  }
}

export default compoundField
