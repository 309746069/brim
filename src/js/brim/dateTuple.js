/* @flow */
import type {DateTuple} from "../lib/TimeWindow"
import brim from "./"

export default function dateTuple(dt: DateTuple) {
  return {
    toSpanArgs() {
      return [brim.time(dt[0]).toTs(), brim.time(dt[1]).toTs()]
    }
  }
}
