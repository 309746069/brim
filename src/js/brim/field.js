/* @flow */

import {isEqual} from "lodash"

import {COMPOUND_FIELD_RGX} from "./compoundField"
import type {FieldData} from "../types/records"
import {withCommas} from "../lib/fmt"
import brim, {type $Field} from "./"

export const ONE_CHAR = 7.39
export const FIELD_PAD = 14
export const PATH_PAD = 12

function field({name, type, value}: FieldData): $Field {
  return {
    name,
    type,
    value,
    queryableValue() {
      if (this.value === null) return "null"
      let WHITE_SPACE = /\s+/
      let COMMA = /,/
      let quote = [WHITE_SPACE, COMMA].some((reg) => reg.test(this.value))
      if (this.type === "string") quote = true
      return quote ? `"${this.value}"` : this.value
    },
    compound() {
      return COMPOUND_FIELD_RGX.test(type)
    },
    toCompound() {
      // $FlowFixMe
      return brim.compoundField(name, type, value)
    },
    toDate() {
      return new Date(+this.value * 1000)
    },
    display() {
      if (value === "(empty)") {
        return ""
      } else if (value === null) {
        return "â¦»"
      } else if (type === "count") {
        return withCommas(value)
      } else if (isEqual(value, {})) {
        return ""
      } else {
        return value
      }
    },
    guessWidth() {
      if (type === "time") {
        return 192
      } else if (name === "_path") {
        return this.display().length * ONE_CHAR + FIELD_PAD + PATH_PAD
      } else if (this.compound()) {
        return this.toCompound().guessWidth()
      } else {
        return Math.ceil(this.display().length * ONE_CHAR + FIELD_PAD)
      }
    }
  }
}

export default field
