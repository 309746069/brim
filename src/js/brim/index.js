/* @flow */
import type {ReturnType} from "../types"
import ast from "./ast"
import compoundField from "./compoundField"
import dateTuple from "./dateTuple"
import field from "./field"
import flatRecordsBuffer from "./flatRecordsBuffer"
import interop from "./interop"
import log from "./log"
import program from "./program"
import randomHash from "./randomHash"
import record from "./record"
import relTime from "./relTime"
import search from "./search"
import space from "./space"
import span from "./span"
import syntax from "./syntax"
import table from "./table"
import time from "./time"

export type $Field = {
  name: string,
  type: string,
  value: string,
  queryableValue: () => string,
  compound: () => boolean,
  toCompound: () => $CompoundField,
  toDate: () => Date,
  display: () => string,
  guessWidth: () => number
}

export type $CompoundField = ReturnType<typeof compoundField>
export type $Search = ReturnType<typeof search>
export type $Log = ReturnType<typeof log>

export type Ts = {
  ns: number,
  sec: number
}

export type Span = [Ts, Ts]

export default {
  table,
  program,
  field,
  compoundField,
  log,
  ast,
  syntax,
  space,
  time,
  relTime,
  span,
  dateTuple,
  flatRecordsBuffer,
  record,
  interop,
  search,
  randomHash
}
