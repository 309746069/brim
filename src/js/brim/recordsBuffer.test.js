/* @flow */
import brim from "./"

let records = [
  {
    id: 0,
    type: [
      {name: "ts", type: "time"},
      {name: "_path", type: "string"},
      {name: "count", type: "count"}
    ],
    values: ["1428917670.000000000", "dns", "2"]
  },
  {id: 0, values: ["1428917670.000000000", "conn", "3"]},
  {id: 0, values: ["1428917670.000000000", "capture_loss", "1"]},
  {id: 0, values: ["1428917640.000000000", "conn", "1"]}
]

test("coverts to an array of records", () => {
  let collector = brim.recordsBuffer()
  collector.add("0", records)
  expect(collector.records()).toMatchSnapshot()
})
