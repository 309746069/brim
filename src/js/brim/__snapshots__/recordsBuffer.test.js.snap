// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coverts to an array of records 1`] = `
Array [
  Array [
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917670.000000000",
    },
    Object {
      "name": "_path",
      "type": "string",
      "value": "dns",
    },
    Object {
      "name": "count",
      "type": "count",
      "value": "2",
    },
  ],
  Array [
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917670.000000000",
    },
    Object {
      "name": "_path",
      "type": "string",
      "value": "conn",
    },
    Object {
      "name": "count",
      "type": "count",
      "value": "3",
    },
  ],
  Array [
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917670.000000000",
    },
    Object {
      "name": "_path",
      "type": "string",
      "value": "capture_loss",
    },
    Object {
      "name": "count",
      "type": "count",
      "value": "1",
    },
  ],
  Array [
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917640.000000000",
    },
    Object {
      "name": "_path",
      "type": "string",
      "value": "conn",
    },
    Object {
      "name": "count",
      "type": "count",
      "value": "1",
    },
  ],
]
`;
